pilot/pkg/xds/sds.go

```diff
@@ -0,0 +1,205 @@
+// Copyright Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package xds
+
+import (
+	"fmt"
+	"strings"
+
+	core "github.com/envoyproxy/go-control-plane/envoy/config/core/v3"
+	tls "github.com/envoyproxy/go-control-plane/envoy/extensions/transport_sockets/tls/v3"
+	"github.com/golang/protobuf/ptypes/any"
+
+	"istio.io/istio/pilot/pkg/model"
+	"istio.io/istio/pilot/pkg/networking/util"
+	"istio.io/istio/pilot/pkg/secrets"
+	"istio.io/istio/pkg/config/schema/gvk"
+)
+
+const (
+	// GatewaySdsCaSuffix is the suffix of the sds resource name for root CA. All resource
+	// names for gateway root certs end with "-cacert".
+	GatewaySdsCaSuffix = "-cacert"
+)
+
+const KubernetesSecretType = "k8s"
+
+type SecretResource struct {
+	Type         string
+	Name         string
+	Namespace    string
+	ResourceName string
+}
+
+func (sr SecretResource) Key() string {
+	return "sds://" + sr.ResourceName
+}
+
+func (sr SecretResource) DependentConfigs() []model.ConfigKey {
+	return []model.ConfigKey{{Kind: gvk.Secret, Name: sr.Name, Namespace: sr.Namespace}}
+}
+
+func (sr SecretResource) Cacheable() bool {
+	return true
+}
+
+var _ model.XdsCacheEntry = SecretResource{}
+
+func parseResourceName(resource, defaultNamespace string) (SecretResource, error) {
+	sep := "/"
+	if strings.HasPrefix(resource, KubernetesSecretType+"://") {
+		res := strings.TrimPrefix(resource, KubernetesSecretType+"://")
+		split := strings.Split(res, sep)
+		namespace := defaultNamespace
+		name := split[0]
+		if len(split) > 1 {
+			namespace = split[0]
+			name = split[1]
+		}
+		return SecretResource{Type: KubernetesSecretType, Name: name, Namespace: namespace, ResourceName: resource}, nil
+	}
+	return SecretResource{}, fmt.Errorf("unknown resource type: %v", resource)
+}
+
+func needsUpdate(proxy *model.Proxy, updates model.XdsUpdates) bool {
+	if proxy.Type != model.Router {
+		return false
+	}
+	if len(updates) == 0 {
+		return true
+	}
+	if len(model.ConfigNamesOfKind(updates, gvk.Secret)) > 0 {
+		return true
+	}
+	return false
+}
+
+// Currently only same namespace is allowed. In the future this will be expanded.
+func (s *SecretGen) proxyAuthorizedForSecret(proxy *model.Proxy, sr SecretResource) bool {
+	// In the initial experimental mode, only istio-system will be allowed. Prior to
+	// being launched, this must be revisited, along with the authorization model for SDS.
+	if proxy.ConfigNamespace != "istio-system" {
```

Comment:
should we use MeshConfig.RootNamespace instead of "istio-system"?

Reply:
Maybe. Since this is a temporary restriction just so we can get users testing this I don't think its a major concern though

Reply:
To double confirm: proxy.ConfigNamespace is verified, and the option to verify is on ? Can we double check ?

Reply:
Yes, we will not start SDS server if auth is not enabled, see https://github.com/istio/istio/pull/26277/files#diff-8cc1e432737fa0d93dc76b5bc10cac83R433